askama_parser/src/expr.rs

```diff
@@ -72,17 +75,74 @@ pub enum Expr<'a> {
 }
 
 impl<'a> Expr<'a> {
-    pub(super) fn arguments(i: &'a str, level: Level) -> ParseResult<'a, Vec<Self>> {
+    pub(super) fn arguments(
+        i: &'a str,
+        level: Level,
+        is_jinja_macro: bool,
+    ) -> ParseResult<'a, Vec<Self>> {
         let (_, level) = level.nest(i)?;
+        let mut named_arguments = HashSet::new();
+        let start = i;
+
         preceded(
             ws(char('(')),
             cut(terminated(
-                separated_list0(char(','), ws(move |i| Self::parse(i, level))),
+                separated_list0(
+                    char(','),
+                    ws(move |i| {
+                        Self::arguments_parsing(
+                            i,
+                            level,
+                            &mut named_arguments,
+                            start,
+                            is_jinja_macro,
+                        )
+                    }),
+                ),
                 char(')'),
             )),
         )(i)
     }
 
+    fn arguments_parsing(
+        i: &'a str,
+        level: Level,
+        named_arguments: &mut HashSet<&'a str>,
+        start: &'a str,
+        is_jinja_macro: bool,
+    ) -> ParseResult<'a, Self> {
+        // Only macros defined in jinja can use named arguments.
+        if is_jinja_macro {
+            alt((
+                move |i| Self::named_argument(i, level, named_arguments, start),
+                move |i| Self::parse(i, level),
+            ))(i)
+        } else {
+            Self::parse(i, level)
+        }
+    }
+
+    fn named_argument(
+        i: &'a str,
+        level: Level,
+        named_arguments: &mut HashSet<&'a str>,
+        start: &'a str,
+    ) -> ParseResult<'a, Self> {
+        let (_, level) = level.nest(i)?;
+        let (i, (argument, _, value)) =
+            tuple((identifier, ws(char('=')), move |i| Self::parse(i, level)))(i)?;
+        if !named_arguments.insert(argument) {
```

Comment:
Please invert this `if` with an early return so that we can avoid rightward drift for the error case.